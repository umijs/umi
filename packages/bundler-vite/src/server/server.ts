import {
  createHttpsServer,
  resolveHttpsConfig,
  createProxy,
} from '@umijs/bundler-utils';
import express from '@umijs/bundler-utils/compiled/express';
import { getDevBanner, logger } from '@umijs/utils';
import http from 'http';
import type {
  DepOptimizationMetadata,
  HmrContext,
  InlineConfig as ViteInlineConfig,
} from '../../compiled/vite';
import { createServer as createViteServer } from '../../compiled/vite';
import type { IConfig } from '../types';
import pluginOnHotUpdate from './plugins/onHotUpdate';

interface IOpts {
  cwd: string;
  port?: number;
  host?: string;
  viteConfig: ViteInlineConfig;
  userConfig: IConfig;
  beforeMiddlewares?: any[];
  afterMiddlewares?: any[];
  /**
   * onDevCompileDone hook
   * @param args  includes 2 fields:
   *              - isFirstCompile:
   *                  it would be true after the dev server is started
   *                  it should be false before each HMR is sent
   *              - stats:
   *                  it would be DepOptimizationMetadata after the dev server is started
   *                  it would be the modules of HMR Context before each HMR is sent
   */
  onDevCompileDone?: (args: {
    time: number;
    isFirstCompile: boolean;
    stats: HmrContext['modules'] | DepOptimizationMetadata;
  }) => Promise<void> | void;
  onBeforeMiddleware?: Function;
}

export async function createServer(opts: IOpts) {
  const startTms = +new Date();
  const { viteConfig, userConfig, onDevCompileDone } = opts;
  const app = express();

  const viteConfigServer = { ...viteConfig.server };

  // 如果启用https 先获取key 和 cert 给vite ws 服务使用
  if (userConfig.https) {
    const httpsConfig = await resolveHttpsConfig(userConfig.https);
    if (httpsConfig) {
      userConfig.https = viteConfigServer.https = {
        key: httpsConfig.key,
        cert: httpsConfig.cert,
      };
    }
  }

  const vite = await createViteServer({
    ...viteConfig,
    // use `handleHotUpdate` vite hook to workaround `onDevCompileDone` umi hook
    ...(typeof onDevCompileDone === 'function'
      ? {
          plugins: viteConfig.plugins!.concat([
            pluginOnHotUpdate(async (modules) => {
              await onDevCompileDone({
                time: 0,
                isFirstCompile: false,
                stats: modules,
              });
            }),
          ]),
        }
      : {}),
    server: { ...viteConfigServer, middlewareMode: true },
  });

  // before middlewares
  opts.beforeMiddlewares?.forEach((m) => app.use(m));

  if (opts.onBeforeMiddleware) {
    opts.onBeforeMiddleware(app);
  }

  // proxy
  if (userConfig.proxy) {
    createProxy(userConfig.proxy, app);
  }

  // after middlewares, insert before vite spaFallbackMiddleware
  // refer: https://github.com/vitejs/vite/blob/2c586165d7bc4b60f8bcf1f3b462b97a72cce58c/packages/vite/src/node/server/index.ts#L508
  if (opts.afterMiddlewares?.length) {
    vite.middlewares.stack.some((s, i) => {
      if ((s.handle as Function).name === 'viteSpaFallbackMiddleware') {
        const afterStacks: typeof vite.middlewares.stack =
          opts.afterMiddlewares!.map((m) => ({
            route: '',
            // TODO: FIXME
            // see: https://github.com/umijs/umi/commit/34d4e4e26a20ff5c7393eab5d3db363cca541379#diff-3a996a9e7a2f94fc8f23c6efed1447eed9567e36ed622bd8547a58e5415087f7R164
            handle: app.use(m.toString().includes(`{ compiler }`) ? m({}) : m),
          }));

        vite.middlewares.stack.splice(i, 0, ...afterStacks);

        return true;
      }

      return false;
    });
  }

  // use vite via middleware way
  app.use(vite.middlewares);

  // writeToDisk(?)
  // mock
  // prerender
  // bundless

  const server = userConfig.https
    ? await createHttpsServer(app, userConfig.https)
    : http.createServer(app);

  if (!server) {
    return null;
  }

  const protocol = userConfig.https ? 'https:' : 'http:';
  const port = opts.port || 8000;

  server.listen(port, async () => {
    if (typeof onDevCompileDone === 'function') {
      await onDevCompileDone({
        time: +new Date() - startTms,
        isFirstCompile: true,
        // @ts-ignore
        stats: vite._optimizeDepsMetadata,
      });
    }

    const banner = getDevBanner(protocol, opts.host, port);

    console.log(banner.before);
    logger.ready(banner.main);
    console.log(banner.after);
  });

  return server;
}
