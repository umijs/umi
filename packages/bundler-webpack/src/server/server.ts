import { createHttpsServer, createProxy } from '@umijs/bundler-utils';
import express from '@umijs/bundler-utils/compiled/express';
import webpack, {
  Configuration,
} from '@umijs/bundler-webpack/compiled/webpack';
import { getDevBanner, lodash, logger } from '@umijs/utils';
import cors from 'cors';
import { createReadStream, existsSync } from 'fs';
import http from 'http';
import { extname, join } from 'path';
import { MESSAGE_TYPE } from '../constants';
import { IConfig } from '../types';
import { createWebSocketServer } from './ws';

interface IOpts {
  cwd: string;
  port?: number;
  host?: string;
  ip?: string;
  webpackConfig: Configuration;
  userConfig: IConfig;
  beforeMiddlewares?: any[];
  afterMiddlewares?: any[];
  onDevCompileDone?: Function;
  onProgress?: Function;
  onBeforeMiddleware?: Function;
}

export async function createServer(opts: IOpts) {
  const { webpackConfig, userConfig } = opts;
  const { proxy } = userConfig;
  const app = express();
  // ws 需要提前初始化
  // 避免在 https 模式下时「Cannot access 'ws' before initialization」的报错
  let ws: ReturnType<typeof createWebSocketServer>;

  // cros
  app.use(
    cors({
      origin: true,
      methods: ['GET', 'HEAD', 'PUT', 'POST', 'PATCH', 'DELETE', 'OPTIONS'],
      credentials: true,
    }),
  );

  // compression
  app.use(require('@umijs/bundler-webpack/compiled/compression')());

  // debug all js file
  app.use((req, res, next) => {
    const file = req.path;
    const filePath = join(opts.cwd, file);
    const ext = extname(filePath);

    if (ext === '.js' && existsSync(filePath)) {
      logger.info(
        '[dev]',
        `${file} is responded with ${filePath}, remove it to use original file`,
      );
      res.sendFile(filePath);
    } else {
      next();
    }
  });

  // before middlewares
  (opts.beforeMiddlewares || []).forEach((m) => app.use(m));

  // Provides the ability to execute custom middleware prior to all other middleware internally within the server.
  if (opts.onBeforeMiddleware) {
    opts.onBeforeMiddleware(app);
  }

  // webpack dev middleware
  const configs = Array.isArray(webpackConfig)
    ? webpackConfig
    : [webpackConfig];
  const progresses: any[] = [];
  if (opts.onProgress) {
    configs.forEach((config) => {
      const progress = {
        percent: 0,
        status: 'waiting',
        details: [],
      };
      progresses.push(progress);
      config.plugins.push(
        new webpack.ProgressPlugin((percent, msg, ...details) => {
          progress.percent = percent;
          progress.status = msg;
          (progress.details as string[]) = details;
          opts.onProgress!({ progresses });
        }),
      );
    });
  }
  const compiler = webpack(configs);

  const webpackDevMiddleware = require('@umijs/bundler-webpack/compiled/webpack-dev-middleware');
  const compilerMiddleware = webpackDevMiddleware(compiler, {
    publicPath: userConfig.publicPath || '/',
    writeToDisk: userConfig.writeToDisk,
    stats: 'none',
    // watchOptions: { ignored }
  });
  app.use(compilerMiddleware);

  // hmr hooks
  let stats: any;
  let isFirstCompile = true;
  compiler.compilers.forEach(addHooks);

  function addHooks(compiler: webpack.Compiler) {
    compiler.hooks.invalid.tap('server', () => {
      sendMessage(MESSAGE_TYPE.invalid);
    });
    compiler.hooks.done.tap('server', (_stats) => {
      stats = _stats;
      sendStats(getStats(stats));
      opts.onDevCompileDone?.({
        stats,
        isFirstCompile,
        time: stats.endTime - stats.startTime,
      });
      isFirstCompile = false;
    });
  }

  function sendStats(
    stats: webpack.StatsCompilation,
    force?: boolean,
    sender?: any,
  ) {
    const shouldEmit =
      !force &&
      stats &&
      (!stats.errors || stats.errors.length === 0) &&
      (!stats.warnings || stats.warnings.length === 0) &&
      stats.assets &&
      stats.assets.every((asset) => !asset.emitted);
    if (shouldEmit) {
      sendMessage(MESSAGE_TYPE.stillOk, null, sender);
      return;
    }
    sendMessage(MESSAGE_TYPE.hash, stats.hash, sender);
    if (
      (stats.errors && stats.errors.length > 0) ||
      (stats.warnings && stats.warnings.length > 0)
    ) {
      if (stats.warnings && stats.warnings.length > 0) {
        sendMessage(MESSAGE_TYPE.warnings, stats.warnings, sender);
      }
      if (stats.errors && stats.errors.length > 0) {
        sendMessage(MESSAGE_TYPE.errors, stats.errors, sender);
      }
    } else {
      sendMessage(MESSAGE_TYPE.ok, null, sender);
    }
  }

  function getStats(stats: webpack.Stats) {
    return stats.toJson({
      all: false,
      hash: true,
      assets: true,
      warnings: true,
      errors: true,
      errorDetails: false,
    });
  }

  function sendMessage(type: string, data?: any, sender?: any) {
    (sender || ws)?.send(JSON.stringify({ type, data }));
  }

  // proxy
  if (proxy) {
    createProxy(proxy, app);
  }

  // after middlewares
  (opts.afterMiddlewares || []).forEach((m) => {
    // TODO: FIXME
    app.use(m.toString().includes(`{ compiler }`) ? m({ compiler }) : m);
  });

  // history fallback
  app.use(
    require('@umijs/bundler-webpack/compiled/connect-history-api-fallback')({
      index: '/',
    }),
  );

  // hmr reconnect ping
  app.use('/__umi_ping', (_, res) => {
    res.end('pong');
  });

  // index.html
  // TODO: remove me
  app.get('/', (_req, res, next) => {
    res.set('Content-Type', 'text/html');
    const htmlPath = join(opts.cwd, 'index.html');
    if (existsSync(htmlPath)) {
      createReadStream(htmlPath).on('error', next).pipe(res);
    } else {
      next();
    }
  });

  let server: http.Server | Awaited<ReturnType<typeof createHttpsServer>>;
  if (userConfig.https) {
    const httpsOpts = userConfig.https;
    if (!httpsOpts.hosts) {
      httpsOpts.hosts = lodash.uniq(
        [
          ...(httpsOpts.hosts || []),
          // always add localhost, 127.0.0.1, ip and host
          '127.0.0.1',
          'localhost',
          opts.ip,
          opts.host !== '0.0.0.0' && opts.host,
        ].filter(Boolean) as string[],
      );
    }
    server = await createHttpsServer(app, httpsOpts);
  } else {
    server = http.createServer(app);
  }
  if (!server) {
    return null;
  }

  ws = createWebSocketServer(server);

  ws.wss.on('connection', (socket) => {
    if (stats) {
      sendStats(getStats(stats), false, socket);
    }
  });

  const protocol = userConfig.https ? 'https:' : 'http:';
  const port = opts.port || 8000;

  server.listen(port, () => {
    const banner = getDevBanner(protocol, opts.host, port);

    console.log(banner.before);
    logger.ready(banner.main);
    console.log(banner.after);
  });

  return server;
}
