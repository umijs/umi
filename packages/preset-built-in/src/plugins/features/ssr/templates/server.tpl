// umi.server.js
import '{{{ RuntimePolyfill }}}';
import { format } from 'url';
import renderServer from '{{{ Renderer }}}';
import { stripBasename, cheerio, handleHTML } from '{{{ Utils }}}';
import { IServerRender } from '@umijs/types';

import { ApplyPluginsType, createMemoryHistory{{ #DynamicImport }}, dynamic{{ /DynamicImport }} } from '{{{ RuntimePath }}}';
{{ #loadingComponent }}
import LoadingComponent from '{{{ loadingComponent }}}';
{{ /loadingComponent }}
import { plugin } from './plugin';
import './pluginRegister';

// origin require module
// https://github.com/webpack/webpack/issues/4175#issuecomment-342931035
const requireFunc = typeof __webpack_require__ === "function" ? __non_webpack_require__ : require;

let routes;

/**
 * server render function
 * @param params
 */
const render: IServerRender = async (params) => {
  let error;
  const {
    origin = '',
    path,
    htmlTemplate = '',
    mountElementId = '{{{ MountElementId }}}',
    context = {},
    mode = '{{{ Mode }}}',
    basename = '{{{ Basename }}}',
    staticMarkup = {{{ StaticMarkup }}},
    forceInitial = {{{ ForceInitial }}},
    removeWindowInitialProps = {{{ RemoveWindowInitialProps }}},
    getInitialPropsCtx,
  } = params;
  let manifest = params.manifest;
  const env = '{{{ Env }}}';

  let html = htmlTemplate || {{{ DEFAULT_HTML_PLACEHOLDER }}};
  let rootContainer = '';
  try {
    // handle basename
    const location = stripBasename(basename, `${origin}${path}`);
    const { pathname } = location;
    // server history
    const history = createMemoryHistory({
      initialEntries: [format(location)],
    });

    /**
     * beforeRenderServer hook, for polyfill global.*
     */
    await plugin.applyPlugins({
      key: 'ssr.beforeRenderServer',
      type: ApplyPluginsType.event,
      args: {
        env,
        path,
        context,
        history,
        mode,
        location,
      },
      async: true,
    });

    /**
     * routes init and patch only once
     * beforeRenderServer must before routes init avoding require error
     */
    if (!routes) {
      // 主要为后面支持按需服务端渲染，单独用 routes 会全编译
      routes = {{{ Routes }}};
      // allow user to extend routes
      plugin.applyPlugins({
        key: 'patchRoutes',
        type: ApplyPluginsType.event,
        args: { routes },
      });
    }

    // for renderServer
    const opts = {
      path,
      history,
      pathname,
      getInitialPropsCtx,
      basename,
      context,
      mode,
      plugin,
      staticMarkup,
      routes,
      isServer: process.env.__IS_SERVER,
    }
    const dynamicImport =  {{{ DynamicImport }}};
    if (dynamicImport && !manifest) {
      try {
        // prerender not work because the manifest generation behind of the prerender
        manifest = requireFunc(`./{{{ ManifestFileName }}}`);
      } catch (_) {}
    }

    // renderServer get rootContainer
    const { pageHTML, pageInitialProps, routesMatched } = await renderServer(opts);
    rootContainer = pageHTML;
    if (html) {
      // plugin for modify html template
      html = await plugin.applyPlugins({
        key: 'ssr.modifyServerHTML',
        type: ApplyPluginsType.modify,
        initialValue: html,
        args: {
          context,
          cheerio,
          routesMatched,
          dynamicImport,
          manifest
        },
        async: true,
      });
      html = await handleHTML({ html, rootContainer, pageInitialProps, mountElementId, mode, forceInitial, removeWindowInitialProps, routesMatched, dynamicImport, manifest });
    }
  } catch (e) {
    // downgrade into csr
    error = e;
  }
  return {
    rootContainer,
    error,
    html,
  }
}

export default render;
