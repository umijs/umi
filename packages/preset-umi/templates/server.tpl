import { getClientRootComponent } from '{{{ serverRendererPath }}}';
import { getRoutes } from './core/route';
import { createHistory as createClientHistory } from './core/history';
import { getPlugins as getClientPlugins } from './core/plugin';
import { PluginManager } from '{{{ umiPluginPath }}}';
import createRequestHandler, { createMarkupGenerator } from '{{{ umiServerPath }}}';

let helmetContext;

try {
  helmetContext = require('./core/helmetContext').context;
} catch { /* means `helmet: false`, do noting */ }

const routesWithServerLoader = {
{{#routesWithServerLoader}}
  '{{{ id }}}': () => import('{{{ path }}}'),
{{/routesWithServerLoader}}
};

export function getPlugins() {
  return getClientPlugins();
}

export function getValidKeys() {
  return [{{#validKeys}}'{{{ . }}}',{{/validKeys}}];
}

export function getManifest() {
  return JSON.parse(require('fs').readFileSync('{{{ assetsPath }}}', 'utf-8'));
}

export function createHistory(opts) {
  return createClientHistory(opts);
}

// TODO: remove global variable
global.g_getAssets = (fileName) => {
  let m = typeof manifest === 'function' ? manifest() : manifest;
  return m.assets[fileName];
};

const manifest = {{{ env }}} === 'development' ? getManifest : getManifest();
const createOpts = {
  routesWithServerLoader,
  PluginManager,
  getPlugins,
  getValidKeys,
  getRoutes,
  manifest,
  getClientRootComponent,
  helmetContext,
  createHistory,
};
const requestHandler = createRequestHandler(createOpts);

export const _markupGenerator = createMarkupGenerator(createOpts);

export default requestHandler;
