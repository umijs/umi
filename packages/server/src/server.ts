import type { RequestHandler } from '@umijs/bundler-utils/compiled/express';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
// import { matchRoutes } from 'react-router-dom';
// import { createServerRoutes } from './routes';
import { normalizeScripts } from './scripts';
import { normalizeStyles } from './styles';

export interface IOpts {
  base: string;
  routes: Record<
    string,
    {
      path: string;
      file: string;
      id: string;
      parentId?: string;
    }
  >;
  links?: Record<string, string>[];
  metas?: Record<string, string>[];
  styles?: (Record<string, string> | string)[];
  favicons?: string[];
  title?: string;
  headScripts?: (Record<string, string> | string)[];
  scripts?: (Record<string, string> | string)[];
  mountElementId?: string;
  esmScript?: boolean;
  modifyHTML?: (html: string, args: { path?: string }) => Promise<string>;
  historyType?: 'hash' | 'browser';
}

export async function getMarkup(
  opts: Omit<IOpts, 'routes'> & {
    path?: string;
  },
) {
  // TODO: use real component
  let markup = ReactDOMServer.renderToString(
    React.createElement('div', { id: opts.mountElementId || 'root' }),
  );

  function propsToString(opts: {
    props: Record<string, any>;
    filters?: string[];
  }) {
    return Object.keys(opts.props)
      .filter((key) => !(opts.filters || []).includes(key))
      .map((key) => `${key}=${JSON.stringify(opts.props[key])}`)
      .join(' ');
  }

  function getScriptContent(script: { src?: string; content?: string }) {
    const attrs = propsToString({
      props: script,
      filters: ['src', 'content'],
    });
    return script.src
      ? `<script${opts.esmScript ? ' type="module"' : ''} ${attrs} src="${
          script.src
        }"></script>`
      : `<script${opts.esmScript ? ' type="module"' : ''} ${attrs}>${
          script.content
        }</script>`;
  }

  function getStyleContent(style: { src?: string; content?: string }) {
    const attrs = propsToString({
      props: style,
      filters: ['src', 'content'],
    });
    return style.src
      ? `<link rel="stylesheet" ${attrs} href="${style.src}" />`
      : `<style ${attrs}>${style.content}</style>`;
  }

  function getTagContent(opts: {
    attrs: Record<string, string>;
    tagName: string;
  }) {
    const attrs = propsToString({
      props: opts.attrs,
    });
    return `<${opts.tagName} ${attrs} />`;
  }

  function withDefaultMetas(metas: IOpts['metas'] = []) {
    const hasAttr = (key: string, value?: string) =>
      metas.some((m) => {
        return value ? m[key]?.toLowerCase() === value.toLowerCase() : m[key];
      });
    return [
      !hasAttr('charset') && { charset: 'utf-8' },
      !hasAttr('name', 'viewport') && {
        name: 'viewport',
        content:
          'width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0',
      },
      !hasAttr('http-equiv', 'X-UA-Compatible') && {
        'http-equiv': 'X-UA-Compatible',
        content: 'ie=edge',
      },
      ...metas,
    ].filter(Boolean) as NonNullable<IOpts['metas']>;
  }

  const favicons: string[] = [];
  if (Array.isArray(opts.favicons)) {
    opts.favicons.forEach((e) => {
      favicons.push(`<link rel="shortcut icon" href="${e}">`);
    });
  }
  const title = opts.title ? `<title>${opts.title}</title>` : '';
  const metas = withDefaultMetas(opts.metas).map((meta) =>
    getTagContent({ attrs: meta, tagName: 'meta' }),
  );
  const links = (opts.links || []).map((link) =>
    getTagContent({ attrs: link, tagName: 'link' }),
  );
  const styles = normalizeStyles(opts.styles || []).map(getStyleContent);
  const headScripts = normalizeScripts(opts.headScripts || []).map(
    getScriptContent,
  );
  const scripts = normalizeScripts(opts.scripts || []).map(getScriptContent);
  markup = [
    `<!DOCTYPE html>
<html>
<head>`,
    metas.join('\n'),
    favicons.join('\n'),
    title,
    links.join('\n'),
    styles.join('\n'),
    headScripts.join('\n'),
    `</head>
<body>`,
    markup,
    scripts.join('\n'),
    `</body>
</html>`,
  ]
    .filter(Boolean)
    .join('\n');
  if (opts.modifyHTML) {
    markup = await opts.modifyHTML(markup, { path: opts.path });
  }
  return markup;
}

export function createRequestHandler(opts: IOpts): RequestHandler {
  return async (req, res, next) => {
    if (
      opts.historyType === 'browser' &&
      opts.base !== '/' &&
      req.path === '/'
    ) {
      // 如果是 browser，并且配置了非 / base，访问 / 时 redirect 到 base 路径
      res.redirect(opts.base);
    } else if (
      req.headers.accept?.includes('text/html') ||
      req.headers.accept === '*/*' ||
      req.path === opts.base
    ) {
      // 匹配路由，不匹配走 next()
      // const routes = createServerRoutes({
      //   routesById: opts.routes,
      // });
      // const matches = matchRoutes(routes, req.path, opts.base);

      // 其他接受 HTML 的请求都兜底返回 HTML
      res.set('Content-Type', 'text/html');
      const markup = await getMarkup({ ...opts, path: req.path });
      res.end(markup);
    } else {
      next();
    }
  };
}
