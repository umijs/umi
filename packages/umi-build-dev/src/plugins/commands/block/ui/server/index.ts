import { join } from 'path';
import { readdirSync } from 'fs';
import { IApi } from 'umi-types';
import Block from './core/Block';
import { Resource } from '../../data';
import { DEFAULT_RESOURCES } from './util';

export default (api: IApi) => {
  // 区块资源可扩展
  let resources: Resource[] = [];

  api.onUISocket(async ({ action, failure, success, send }) => {
    if (!resources.length) {
      resources = api.applyPlugins('addBlockUIResource', {
        initialValue: DEFAULT_RESOURCES,
      });
      resources = api.applyPlugins('modifyBlockUIResources', {
        initialValue: resources,
      });
    }

    const blockService = new Block(api);
    blockService.init(send);
    const { type, payload = {}, lang } = action;

    const dir = join(__dirname, 'socketHandlers');
    const files = readdirSync(dir)
      .filter(f => f.charAt(0) !== '.')
      .map(f => f.replace(/\.(js|ts)$/, ''));

    if (files.includes(type)) {
      try {
        const fn = require(join(dir, type)).default;
        await fn({
          api,
          success,
          failure,
          send,
          payload,
          lang,
          blockService,
          resources,
        });
      } catch (e) {
        console.error(e);
        failure({
          message: e.message,
          success: false,
        });
      }
    }
  });
};
