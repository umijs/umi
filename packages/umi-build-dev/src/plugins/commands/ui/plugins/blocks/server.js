export default function(api) {
  function getBlocks() {
    return [
      {
        name: 'demo',
      },
    ];
    return api.service.userConfig.getConfig({ force: true });
  }

  api.onSocketData((type, payload, { send }) => {
    console.log(`[LOG] ${type} ${JSON.stringify(payload)}`);

    switch (type) {
      case 'blocks':
        console.log('run command: block', payload);
        api.service
          .runCommand('block', {
            _: payload,
          })
          .then(() => {
            console.log('blocks done');
            send('blocks/save', getConfig());
          });
        break;
      case 'blocks/fetch':
        send('blocks/save', getBlocks());
        break;
      default:
        break;
    }
  });
}
