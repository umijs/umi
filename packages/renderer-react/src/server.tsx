import React from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { StaticRouter } from 'react-router-dom/server';
import { AppContext } from './appContext';
import { Routes } from './browser';
import { createClientRoutes } from './routes';
import { IRouteComponents, IRoutesById } from './types';

// store helmet data
const helmetContext = {};

// get helmet context
export function getHelmetContext() {
  return helmetContext;
}

// Get the root React component for ReactDOMServer.renderToString
export async function getClientRootComponent(opts: {
  routes: IRoutesById;
  routeComponents: IRouteComponents;
  pluginManager: any;
  location: string;
  loaderData: { [routeKey: string]: any };
  manifest: any;
}) {
  const basename = '/';
  const components = { ...opts.routeComponents };
  const clientRoutes = createClientRoutes({
    routesById: opts.routes,
    routeComponents: components,
  });
  let rootContainer = (
    <StaticRouter basename={basename} location={opts.location}>
      <Routes />
    </StaticRouter>
  );
  for (const key of [
    // Lowest to the highest priority
    'innerProvider',
    'i18nProvider',
    'accessProvider',
    'dataflowProvider',
    'outerProvider',
    'rootContainer',
  ]) {
    rootContainer = opts.pluginManager.applyPlugins({
      type: 'modify',
      key: key,
      initialValue: rootContainer,
      args: {},
    });
  }
  return (
    <Html loaderData={opts.loaderData} manifest={opts.manifest}>
      <HelmetProvider context={helmetContext}>
        <AppContext.Provider
          value={{
            routes: opts.routes,
            routeComponents: opts.routeComponents,
            clientRoutes,
            pluginManager: opts.pluginManager,
            basename,
            clientLoaderData: {},
            serverLoaderData: opts.loaderData,
          }}
        >
          {rootContainer}
        </AppContext.Provider>
      </HelmetProvider>
    </Html>
  );
}

function Html({ children, loaderData, manifest }: any) {
  // TODO: 处理 head 标签，比如 favicon.ico 的一致性
  // TODO: root 支持配置
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        {manifest.assets['umi.css'] && (
          <link rel="stylesheet" href={manifest.assets['umi.css']} />
        )}
      </head>
      <body>
        <noscript
          dangerouslySetInnerHTML={{
            __html: `<b>Enable JavaScript to run this app.</b>`,
          }}
        />
        <div id="root">{children}</div>
        <script
          dangerouslySetInnerHTML={{
            __html: `window.__UMI_LOADER_DATA__ = ${JSON.stringify(
              loaderData,
            )}`,
          }}
        />
      </body>
    </html>
  );
}
