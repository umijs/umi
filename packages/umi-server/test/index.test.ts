import { join } from 'path';
import { readdirSync } from 'fs';
import { fork } from 'child_process';
import { winPath } from 'umi-utils';
import server from '..';

async function build({ cwd }) {
  return new Promise((resolve, reject) => {
    const umiPath = join(winPath(__dirname), '..', '..', 'umi', 'bin', 'umi');
    const env = {
      UMI_UI: 'none',
      COMPRESS: 'none',
      PROGRESS: 'none',
      COVERAGE: 1,
    };
    const child = fork(umiPath, ['build'], {
      cwd,
      env,
    });
    child.on('exit', code => {
      if (code === 1) {
        reject(new Error('Build failed'));
        process.exit(code);
      } else {
        resolve();
      }
    });
  });
}

const fixtures = join(winPath(__dirname), 'fixtures');

describe('build', () => {
  beforeAll(async () => {
    const dirs = readdirSync(fixtures).filter(dir => dir.charAt(0) !== '.');

    const buildPromise = dirs.map(dir => build({ cwd: join(fixtures, dir) }));
    await Promise.all(buildPromise);
  });

  afterAll(done => {
    done();
  });

  it('ssr', async () => {
    const render = server({
      root: join(fixtures, 'ssr', 'dist'),
      publicPath: '/',
    });
    const { ssrHtml } = await render({
      req: {
        url: '/',
      },
    });
    expect(ssrHtml).toMatch(/Hello UmiJS SSR/);
  });

  it('ssr commonjs require', async () => {
    const serverCjs = require('..');
    const render = serverCjs({
      root: join(fixtures, 'ssr', 'dist'),
      publicPath: '/',
    });
    const { ssrHtml } = await render({
      req: {
        url: '/',
      },
    });
    expect(ssrHtml).toMatch(/Hello UmiJS SSR/);
  });

  it('ssr-styles', async () => {
    const render = server({
      root: join(fixtures, 'ssr-styles', 'dist'),
      publicPath: '/',
    });

    const { ssrHtml: indexHtml } = await render({
      req: {
        url: '/',
      },
    });

    expect(indexHtml).toMatchSnapshot();

    const { ssrHtml: newsHtml } = await render({
      req: {
        url: '/news',
      },
    });
    expect(newsHtml).toMatchSnapshot();
  });

  it('ssr-dynamicImport', async () => {
    const render = server({
      root: join(fixtures, 'ssr-dynamicImport', 'dist'),
      publicPath: '/',
    });
    const { ssrHtml, chunkMap } = await render({
      req: {
        url: '/',
      },
    });
    expect(ssrHtml).toMatchSnapshot();
    expect(chunkMap).toEqual({
      css: [],
      js: ['umi.js', 'p__index.async.js'],
    });

    const renderPostProcessHtml = server({
      root: join(fixtures, 'ssr-dynamicImport', 'dist'),
      publicPath: '/',
      postProcessHtml: (html, { load }) => {
        const $ = load(html);
        $('html').attr('lang', 'zh');
        return $.html();
      },
    });
    const { ssrHtml: ssrHtmlPostProcessHtml } = await renderPostProcessHtml({
      req: {
        url: '/',
      },
    });
    expect(ssrHtmlPostProcessHtml).toMatchSnapshot();
  });
});
