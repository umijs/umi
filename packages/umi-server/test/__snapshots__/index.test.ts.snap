// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build ssr-dynamicImport 1`] = `
"<!DOCTYPE html><html data-reactroot><head><link rel=\\"stylesheet\\" href=\\"/\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
  window.g_initialData = {\\"list\\":[{\\"name\\":\\"Alice\\"},{\\"name\\":\\"Jack\\"},{\\"name\\":\\"Tony\\"}]};</script><script>window.routerBase = \\"/\\";</script><link rel=\\"preload\\" href=\\"/p__index.async.js\\" as=\\"script\\"></head><body><div id=\\"root\\"><div class=\\"wrapper\\"><h1>Hello UmiJS SSR</h1><ul><li>Alice</li><li>Jack</li><li>Tony</li></ul><button>0</button></div></div><script src=\\"/umi.js\\"></script></body></html>"
`;

exports[`build ssr-dynamicImport 2`] = `
"<!DOCTYPE html><html data-reactroot lang=\\"zh\\"><head><link rel=\\"stylesheet\\" href=\\"/\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
  window.g_initialData = {\\"list\\":[{\\"name\\":\\"Alice\\"},{\\"name\\":\\"Jack\\"},{\\"name\\":\\"Tony\\"}]};</script><script>window.routerBase = \\"/\\";</script><link rel=\\"preload\\" href=\\"/p__index.async.js\\" as=\\"script\\"></head><body><div id=\\"root\\"><div class=\\"wrapper\\"><h1>Hello UmiJS SSR</h1><ul><li>Alice</li><li>Jack</li><li>Tony</li></ul><button>0</button></div></div><script src=\\"/umi.js\\"></script></body></html>"
`;

exports[`build ssr-styles 1`] = `
"<!DOCTYPE html><html data-reactroot><head><link rel=\\"stylesheet\\" href=\\"/umi.css\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
  window.g_initialData = {\\"list\\":[{\\"name\\":\\"Alice\\"},{\\"name\\":\\"Jack\\"},{\\"name\\":\\"Tony\\"}]};</script><script>window.routerBase = \\"/\\";</script><link rel=\\"stylesheet\\" href=\\"/umi.css\\"></head><body><div id=\\"root\\"><div class=\\"wrapper\\"><h1>Hello UmiJS SSR Styles</h1><ul><li>Alice</li><li>Jack</li><li>Tony</li></ul><button>0</button></div></div><script src=\\"/umi.js\\"></script></body></html>"
`;

exports[`build ssr-styles 2`] = `
"<!DOCTYPE html><html data-reactroot><head><link rel=\\"stylesheet\\" href=\\"/umi.css\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
  window.g_initialData = {};</script><script>window.routerBase = \\"/\\";</script><link rel=\\"stylesheet\\" href=\\"/umi.css\\"></head><body><div id=\\"root\\"><div class=\\"news\\"><h1>Hello UmiJS SSR Styles</h1></div></div><script src=\\"/umi.js\\"></script></body></html>"
`;
